File: data_structures/c++/doubly_linked_list.cpp
#include <bits/stdc++.h>
using namespace std;

struct node
{
        int data;
        node *next;
        node *previous;
};
struct node *head = NULL;

void display()
{
        struct node *temp;
        if (head == NULL)
        {
                cout << "list is empty\n\n";
        }
        else
        {
                temp = head;
                cout << "Doubly linked list: ";
                while (temp != NULL)
                {
                        cout << temp->data << " ";
                        temp = temp->next;
                }
        }
        cout << "\n\n";
}

void insert_beg(int x)
{
        struct node *new_node = (struct node *)malloc(sizeof(struct node));

        if (head == NULL)
        {
                new_node->data = x;
                new_node->previous = NULL;
                new_node->next = NULL;
                head = new_node;
        }
        else
        {
                new_node->data = x;
                new_node->next = head;
                new_node->previous = NULL;
                head->previous = new_node;
                head = new_node;
        }
        display();
}
void insert_end(int x)
{
        struct node *new_node = (struct node *)malloc(sizeof(struct node));
        if (head == NULL)
        {
                new_node->data = x;
                new_node->next = NULL;
                new_node->previous = NULL;
                head = new_node;
        }
        else
        {
                new_node->data = x;
                new_node->next = NULL;

                struct node *temp = head;

                while (temp->next != NULL)
                {
                        temp = temp->next;
                }
                new_node->previous = temp;
                temp->next = new_node;
        }
        display();
}
void insert_at_n(int x, int n)
{
        struct node *new_node = (struct node *)malloc(sizeof(struct node));
        if (head == NULL && n == 0)
        {
                new_node->data = x;
                new_node->next = NULL;
                new_node->previous = NULL;
                head = new_node;
        }
        else if (head == NULL && n > 0)
        {
                cout << "Invalid Index" << endl;
        }
        else if (n == 0)
        {
                insert_beg(x);
        }
        else
        {
                struct node *temp = head;
                bool valid = true;

                while (--n)
                {
                        temp = temp->next;
                        if (temp == NULL)
                        {
                                cout << "Invalid Index" << endl;
                                valid = false;
                                break;
                        }
                }
                if (valid)
                {
                        if (temp->next == NULL)
                        {
                                new_node->previous = temp;
                                temp->next = new_node;
                        }
                        else
                        {
                                new_node->data = x;
                                new_node->next = temp->next;
                                temp->next->previous = new_node;
                                temp->next = new_node;
                                new_node->previous = temp;
                        }
                }
                display();
        }
}
void delete_from_beg()
{
        if (head == NULL)
        {
                cout << "List is already empty!" << endl;
        }
        else
        {
                struct node *temp = head;
                head = head->next;
                temp->next = NULL;
                head->previous = NULL;
        }
        display();
}
void delete_from_end()
{
        if (head == NULL)
        {
                cout << "List is already empty!" << endl;
        }
        else
        {
                struct node *temp = head;
                while (temp->next->next != NULL)
                {
                        temp = temp->next;
                }
                temp->next = NULL;
        }
        display();
}
void delete_from_n(int n)
{
        if (head == NULL)
        {
                cout << "List is already empty!" << endl;
        }
        else if (n == 0)
        {
                struct node *temp = head;
                head = head->next;
                temp->next = NULL;
                head->previous = NULL;
        }
        else
        {
                struct node *temp = head;
                bool valid = true;
                while (--n)
                {
                        temp = temp->next;
                        if (temp == NULL)
                        {
                                cout << "Invalid Index" << endl;
                                valid = false;
                                break;
                        }
                }
                if (valid)
                {
                        if (temp->next->next != NULL)
                        {
                                temp->next->next->previous = temp;
                        }
                        temp->next = temp->next->next;
                }
        }
        display();
}
void delete_the_element(int x)
{
        if (head == NULL)
        {
                cout << "List is already empty!" << endl;
        }
        else
        {
                bool found = false;
                int n = 0;
                struct node *temp = head;
                while (temp != NULL)
                {
                        //cout<<temp->data<<"  ";
                        if (temp->data == x)
                        {
                                delete_from_n(n);
                                n--;
                                found = true;
                        }
                        temp = temp->next;
                        n++;
                }
                if (!found)
                {
                        cout << "The element " << x << " is not present in list" << endl;
                }
        }
}
void custom_display()
{
        struct node *temp;
        if (head == NULL)
        {
                cout << "list is empty\n\n";
        }
        else
        {
                temp = head;
                cout << "|" << head << "|-->";
                while (temp != NULL)
                {
                        cout << "|" << temp->previous << "-" << temp->data << "-" << temp->next << "|";
                        if (temp->next != NULL)
                        {
                                cout << "<-->";
                        }
                        temp = temp->next;
                }
        }
        cout << "\n\n";
}
void reverse_display()
{
        if (head == NULL)
        {
                cout << "List is empty!" << endl;
        }
        else
        {
                struct node *temp = head;
                while (temp->next != NULL)
                {
                        temp = temp->next;
                }
                cout << "reverse: ";
                while (temp != NULL)
                {
                        cout << temp->data << " ";
                        temp = temp->previous;
                }
                cout << endl;
        }
}

int main()
{
        insert_beg(2);
        insert_beg(7);
        insert_beg(5);
        insert_beg(4);
        insert_end(45);
        insert_at_n(78, 4);
        delete_from_end();
        delete_from_n(2);
        delete_from_beg();
        reverse_display();
        custom_display();
}

File: data_structures/doubly_linked_list.py
import sys
def stdwrite(line): return sys.stdout.write(line)


class DoublyLinkedList:
    class __Node:
        def __init__(self):
            self.previous = None
            self.data = None
            self.next = None

        def set_data(self, data):
            self.data = data

        def get_data(self):
            return self.data

        def set_next(self, next):
            self.next = next

        def get_next(self):
            return self.next

        def get_previous(self):
            return self.previous
        
        def set_previous(self, previous):
            self.previous = previous
    __head = None
    def insert_at_beg(self, data):
        if not DoublyLinkedList.__head:
            new = self.__Node()
            new.set_data(data)
            DoublyLinkedList.__head = new
            self.display()
        else:
            new = self.__Node()
            new.set_data(data)
            new.set_next(DoublyLinkedList.__head)
            DoublyLinkedList.__head.set_previous(new)
            DoublyLinkedList.__head = new
            self.display()

    def insert_at_end(self, data):
        if not DoublyLinkedList.__head:
            new = self.__Node()
            new.set_data(data)
            DoublyLinkedList.__head = new
            self.display()
        else:
            new = self.__Node()
            new.set_data(data)
            temp = DoublyLinkedList.__head
            while temp.get_next():
                temp = temp.get_next()
            new.set_previous(temp)
            temp.set_next(new)
            self.display()

    def insert_at_n(self, data, n):
        if not DoublyLinkedList.__head:
            new = self.__Node()
            new.set_data(data)
            DoublyLinkedList.__head = new
            self.display()
        elif n==0:
            self.insert_at_beg(data)
        else:
            counter = 0
            temp = DoublyLinkedList.__head
            while temp:
                if counter==n-1:
                    if not temp.get_next():
                        self.insert_at_end(data)
                        return True
                    new = self.__Node()
                    new.set_data(data)
                    new.set_next(temp.get_next())
                    new.set_previous(temp.get_next().get_previous())
                    temp.get_next().set_previous(new)
                    temp.set_next(new)
                    self.display()
                    return True
                temp = temp.get_next()
                counter += 1
            stdwrite("This is index does not exist in the list\n")
            return False

    def delete_from_beg(self):
        if not DoublyLinkedList.__head:
            stdwrite("List is already empty\n")
        else:
            DoublyLinkedList.__head = DoublyLinkedList.__head.get_next()
            self.display()

    def delete_from_end(self):
        if not DoublyLinkedList.__head:
            stdwrite("List is already empty\n")
        else:
            temp = DoublyLinkedList.__head
            if not temp.get_next():
                DoublyLinkedList.__head = None
            else:
                while temp.get_next().get_next():
                    temp = temp.get_next()
                temp.set_next(None)
                self.display()

    def delete_from_n(self, n):
        if not DoublyLinkedList.__head:
            stdwrite("List is already empty\n")
        elif n == 0:
            self.delete_from_beg()
        else:
            temp = DoublyLinkedList.__head
            counter = 0
            while temp:
                if counter == n-1:
                    try: 
                        if not temp.get_next().get_next():
                            self.delete_from_end()
                            return True
                        temp.get_next().get_next().set_previous(temp)
                        temp.set_next(temp.get_next().get_next())
                        self.display()
                        return True
                    except:
                        pass
                temp = temp.get_next()
                counter += 1
            stdwrite("This position does not exist\n")

    def display(self):
        if not DoublyLinkedList.__head:
            stdwrite("List is empty\n")
        else:
            stdwrite("Doubly Linked List: head")
            temp = DoublyLinkedList.__head
            while temp:
                stdwrite(" <-> "+str(temp.get_data()))
                temp = temp.get_next()
            stdwrite("\n")

    def __reverse(self, temp):
        if temp == None:
            return
        self.__reverse(temp.get_next())
        stdwrite("<- "+str(temp.get_data())+" ")

    def print_reverse(self):
        if not DoublyLinkedList.__head:
            stdwrite("The list is empty\n")
        else:
            stdwrite("Reverse Linked List: ")
            temp = DoublyLinkedList.__head
            self.__reverse(temp)
            stdwrite("<-> head")


if __name__=="__main__":
    li = DoublyLinkedList()
    li.insert_at_beg(4)
    li.insert_at_beg(5)
    li.insert_at_beg(41)
    li.insert_at_end(7)
    li.insert_at_n(10, 1)
    li.delete_from_beg()
    li.delete_from_end()
    li.delete_from_n(2)
    li.print_reverse()




